### ISSUE TITLE
Odd bug in openmc.Mgxs/openmc.Library

### ISSUE THREAD

**Issue by nelsonag** [mit-crpg/openmc#657]
When trying to generate a new MGXS test, I found a very odd issue with the openmc.Library class.  I want to say it's related to confusion in the case of both `scatter matrix` and `transport` mgxs types being present in a library.  To recreate, go to test_mgxs_library_no_nuclides (though I think any of the MGXS tests will do it), and replace the current self.mgxs_types declaration on line 28 with the following:
`self.mgxs_lib.mgxs_types = ['total', 'transport', 'nu-transport',
                                    'absorption', 'capture', 'fission',
                                    'nu-fission', 'kappa-fission',
                                    'scatter matrix']`

Yes, it has to be that (though maybe when we learn more, the pattern will emerge).  If you take away `scatter matrix`, the bug wont show up. Take away a mgxs_type from before it, and it too, wont show up.

Ok, now run the test.  This is what you get when the code is trying to calculate either transport or nu-transport's cross section.  
```
Traceback (most recent call last):
  File "test_mgxs_library_no_nuclides.py", line 85, in <module>
    harness.main()
  File "../testing_harness.py", line 258, in main
    self.execute_test()
  File "../testing_harness.py", line 269, in execute_test
    results = self._get_results()
  File "test_mgxs_library_no_nuclides.py", line 65, in _get_results
    df = mgxs.get_pandas_dataframe()
  File "../../openmc/mgxs/mgxs.py", line 1467, in get_pandas_dataframe
    df = self.xs_tally.get_pandas_dataframe(
  File "../../openmc/mgxs/mgxs.py", line 307, in xs_tally
    self._xs_tally = self.rxn_rate_tally / self.tallies['flux']
  File "../../openmc/mgxs/mgxs.py", line 2131, in rxn_rate_tally
    self.tallies['total'] - self.tallies['scatter-1']
  File "../../openmc/tallies.py", line 2554, in __sub__
    new_tally = self.hybrid_product(other, binary_op='-')
  File "../../openmc/tallies.py", line 1978, in hybrid_product
    new_tally._mean = data['self']['mean'] - data['other']['mean']
ValueError: operands could not be broadcast together with shapes (2,1,1) (4,1,1)
```

As you can see, the code cant perform the subtraction operation in `total - scatter-1` since `scatter-1` has an extra dimension.  When you print it out you see its actually a scatter-1 2x2 matrix.  This makes me think the code is confusing the scatter-1 for transport with the scatter-p1 for the scatter matrix.  This is corroborated by the fact that if I set the legendre_order to 0, it goes away.

You may be asking yourself, why do you need all those extra scores to get that behavior to show up? Only a scatter matrix and transport should be needed right?  I have no idea why it doesnt show up in that case. Maybe once this riddle is solved we will know more.

**Comment by nelsonag** [comment]
Also, should the filter type change on mgxs.py:2129 be changed back to `energyout`? This doesnt fix the bug, but it is kind of weird to change the filter type and not change it back in case something else needs it in its prior state.

**Comment by wbinventor** [comment]
This is certainly an odd bug @nelsonag! I think I know what the problem is, though I haven't tried to reproduce the situation yet and won't have time to do so for a few days. The (`MGXS.load_from_statepoint(...)`)[https://github.com/mit-crpg/openmc/blob/develop/openmc/mgxs/mgxs.py#L694] method uses the `StatePoint.get_tally(...)` method to "find" the `Tally` objects in the `StatePoint` with a certain set of filters, nuclides and scores. The `get_tally(...)` method looks for the *first* `Tally` which contains *all* of the filters, scores, and nuclides. The caveat is that the `Tally` *may* contain more filters, nuclides and/or scores than those provided in the input parameters - *i.e.*, you only have to provide a subset of the specification defining at `Tally` to retrieve it. In this case, the specification of filters, scores and nuclides for the `scatter-1` tally in the `TransportXS` is met by the `scatter-p1` tally for the `ScatterMatrixXS` which has an additional energyout filter.

This behavior for the `StatePoint.get_tally(...)` method is intentional and can be convenient if you only have a few `Tally` objects, but clearly is dangerous if there are many `Tally` objects with similar specifications. I think the solution would entail either adding a new boolean `exact` parameter to `StatePoint.get_tally(...)` to specify that the input parameters exactly describe the queried `Tally` object,  or simply making `StatePoint.get_tally(...)` more strict and only returning a `Tally` if it meets the input parameters but nothing more.

**Comment by paulromano** [comment]
I find it rather useful that `get_tally` is not strict so I would vote for an `exact` option or the like.

**Comment by nelsonag** [comment]
Closed with #660
